Classes/Capture/AddNoteViewController.swift

```diff
@@ -54,12 +54,114 @@ final class AddNoteViewController: UIViewController {
         fatalError("init(coder:) has not been implemented")
     }
 
+    private lazy var toolsBar: UIToolbar = {
+        let bar = UIToolbar()
+        let button1 = UIButton( type: .custom )
+        button1.setImage( UIImage( named: "linkkey.png" ), for: .normal )
+        button1.addTarget(self, action: #selector(addLinkMarkup), for: .touchUpInside)
+        button1.frame = CGRect( x: 0, y: 0, width: 53, height: 51 )
+        button1.bounds = CGRect( x: 0, y: 0, width: 53, height: 51 )
+        let urlButton = UIBarButtonItem( customView: button1 )
+        let button2 = UIButton( type: .custom )
+        button2.setImage( UIImage( named: "datepick.png" ), for: .normal )
+        button2.addTarget(self, action: #selector(setDatePickerView), for: .touchUpInside)
+        button2.frame = CGRect( x: 0, y: 0, width: 53, height: 51 )
+        button2.bounds = CGRect( x: 0, y: 0, width: 53, height: 51 )
+        let dateButton = UIBarButtonItem( customView: button2 )
+        let spacer = UIBarButtonItem(barButtonSystemItem: .flexibleSpace, target: nil, action: nil)
+        bar.items = [urlButton, dateButton, spacer ]
+        bar.sizeToFit()
+        return bar
+    }()
+
+    @objc func addLinkMarkup() {
+        guard self.textView.text != nil else { return } // If no selected text, ignore.  Consider popup?
+        if let range = self.textView.selectedTextRange, !range.isEmpty {
+            let selectedText = self.textView.text( in: range )
+            self.textView.replace( range, withText: "[[\(selectedText!)][]]" )
+            let newCursor = self.textView.position(from: range.end, offset: 4)!
```

Comment:
Also, let's avoid force unwrapping here as well. Simple `if` statement will do, something like this:
```swift
if let newCursor = self.textView.position(from: range.end, offset: 4) {
    self.textView.selectedTextRange = self.textView.textRange(from: newCursor, to: newCursor)
}
```

P.S. Force-unwraps are leading to the application crash if the value cannot be extracted; that's why it is better to protect users. In the worst case for this scenario, we will keep the cursor where it was but will not crash and a user will not lose the data in the input field.

Reply:
Thanks.  I realize that force unwraps *could* lead to a crash, but I reasoned that the offset had to be good because I'd just inserted text beyond that point.  I agree using a guard is cleaner though.  I suppose the inserted text could have failed if we went beyond the allowed size of the note or something.